stream.go

```diff
@@ -461,74 +520,58 @@ func (cs *clientStream) RecvMsg(m interface{}) (err error) {
 
 	// Special handling for non-server-stream rpcs.
 	// This recv expects EOF or errors, so we don't collect inPayload.
-	err = recv(cs.p, cs.codec, cs.s, cs.dc, m, *cs.c.maxReceiveMessageSize, nil, cs.decomp)
+	err = recv(a.p, cs.codec, a.s, a.dc, m, *cs.c.maxReceiveMessageSize, nil, a.decomp)
 	if err == nil {
 		return toRPCErr(errors.New("grpc: client streaming protocol violation: get <nil>, want <EOF>"))
 	}
 	if err == io.EOF {
-		return cs.s.Status().Err() // non-server streaming Recv returns nil on success
+		return a.s.Status().Err() // non-server streaming Recv returns nil on success
 	}
 	return toRPCErr(err)
 }
 
-func (cs *clientStream) CloseSend() error {
+func (a *csAttempt) closeSend() {
+	cs := a.cs
 	if cs.sentLast {
-		return nil
+		return
 	}
 	cs.sentLast = true
-	cs.t.Write(cs.s, nil, nil, &transport.Options{Last: true})
-	// We ignore errors from Write and always return nil here.  Any error it
-	// would return would also be returned by a subsequent RecvMsg call, and the
-	// user is supposed to always finish the stream by calling RecvMsg until it
-	// returns err != nil.
-	return nil
+	cs.attempt.t.Write(cs.attempt.s, nil, nil, &transport.Options{Last: true})
+	// We ignore errors from Write.  Any error it would return would also be
+	// returned by a subsequent RecvMsg call, and the user is supposed to always
+	// finish the stream by calling RecvMsg until it returns err != nil.
 }
 
-func (cs *clientStream) finish(err error) {
-	if err == io.EOF {
-		// Ending a stream with EOF indicates a success.
-		err = nil
-	}
-	cs.mu.Lock()
-	defer cs.mu.Unlock()
-	if cs.finished {
-		return
-	}
-	cs.finished = true
-	cs.t.CloseStream(cs.s, err)
-	for _, o := range cs.opts {
-		o.after(cs.c)
-	}
-	if cs.done != nil {
-		cs.done(balancer.DoneInfo{
+func (a *csAttempt) finish(err error) {
+	a.mu.Lock()
+	a.t.CloseStream(a.s, err)
+
+	if a.done != nil {
+		a.done(balancer.DoneInfo{
 			Err:           err,
 			BytesSent:     true,
-			BytesReceived: cs.s.BytesReceived(),
+			BytesReceived: a.s.BytesReceived(),
 		})
-		cs.done = nil
 	}
-	if cs.statsHandler != nil {
+	if a.statsHandler != nil {
 		end := &stats.End{
 			Client:  true,
 			EndTime: time.Now(),
 			Error:   err,
 		}
-		cs.statsHandler.HandleRPC(cs.statsCtx, end)
-	}
-	cs.cancel()
-	if !cs.tracing {
-		return
+		a.statsHandler.HandleRPC(a.ctx, end)
 	}
-	if cs.trInfo.tr != nil {
+	if a.trInfo.tr != nil {
```

Comment:
Missing the outter conditional?
```
if EnableTracing {
  if a.trInfo.tr != nil {
  ....
}
```

Reply:
We are holding a.mu here, so looking at only `a.trInfo.tr` is good.  The only reason to check `EnableTracing` in the other cases is to avoid taking the lock if tracing is disabled globally.